app/controllers/pieces_controller.rb

```diff
@@ -17,6 +21,10 @@ def show
 
   def edit
     @piece = Piece.find(params[:id])
+    if current_user != @piece.user && !current_user.admin
+      flash[:notice] = "You can only edit a piece you created"
+      redirect_to piece_path(@piece)
+    end
```

Comment:
Abstract out the "can edit?" logic here into the user model itself:

*pieces_controller.rb*
```
class PiecesController < ApplicationController
  ..
  def edit
    @piece = Piece.find(params[:id])
    unless current_user.can_edit?(@piece)
      flash[:notice] = "You can only edit ..."
      redirect_to piece_path(@piece)
    end
  end

  ...
end
```

*user.rb*
```
class User < ActiveRecord::Base
  ...
  def can_edit?(piece)
    self.admin || self == piece.user
  end
end
```